unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -1725,6 +1731,77 @@ inputDescription input =
       -- just trying to recover the syntax the user wrote
       These path _branch -> pure (Path.toText' path)
 
+handleStructuredFindI :: HQ.HashQualified Name -> Cli ()
+handleStructuredFindI rule = do
+  Cli.Env {codebase} <- ask
+  (ppe, names, rules0) <- lookupRewrite InvalidStructuredFind rule
+  let rules = snd <$> rules0
+  let fqppe = PPED.unsuffixifiedPPE ppe
+  results :: [(HQ.HashQualified Name, Referent)] <- pure $ do
+    r <- Set.toList (Relation.ran $ Names.terms (NamesWithHistory.currentNames names))
+    Just hq <- [PPE.terms fqppe r]
+    fullName <- [HQ'.toName hq]
+    guard (not (Name.beginsWithSegment fullName Name.libSegment))
+    Referent.Ref _ <- pure r
+    Just shortName <- [PPE.terms (PPED.suffixifiedPPE ppe) r]
+    pure (HQ'.toHQ shortName, r)
+  let ok t@(_, Referent.Ref (Reference.DerivedId r)) = do
+        oe <- Cli.runTransaction (Codebase.getTerm codebase r)
+        pure $ (t, maybe False (\e -> any ($ e) rules) oe)
+      ok t = pure (t, False)
+  results0 <- traverse ok results
+  let results = [(hq, r) | ((hq, r), True) <- results0]
+  let toNumArgs = Text.unpack . Reference.toText . Referent.toReference . view _2
+  #numberedArgs .= map toNumArgs results
+  Cli.respond (ListStructuredFind (fst <$> results))
+
+lookupRewrite :: (HQ.HashQualified Name -> Output) -> HQ.HashQualified Name -> Cli (PPED.PrettyPrintEnvDecl, NamesWithHistory, [(Term Symbol Ann -> Maybe (Term Symbol Ann), Term Symbol Ann -> Bool)])
+lookupRewrite onErr rule = do
+  Cli.Env {codebase} <- ask
+  root <- Cli.getRootBranch
+  currentBranch <- Cli.getCurrentBranch0
+  currentPath' <- Cli.getCurrentPath
+  hqLength <- Cli.runTransaction Codebase.hashLength
+  let currentNames = NamesWithHistory.fromCurrentNames $ Branch.toNames currentBranch
+  let ppe = Backend.getCurrentPrettyNames hqLength (Backend.WithinStrict (Path.unabsolute currentPath')) root
+  ot <- Cli.getTermFromLatestParsedFile rule
+  ot <- case ot of
+    Just _ -> pure ot
+    Nothing -> do
+      case NamesWithHistory.lookupHQTerm rule currentNames of
+        s
+          | Set.size s == 1,
+            Referent.Ref (Reference.DerivedId r) <- Set.findMin s ->
+              Cli.runTransaction (Codebase.getTerm codebase r)
+        s -> Cli.returnEarly (TermAmbiguous (PPE.suffixifiedPPE ppe) rule s)
```

Comment:
Is it possible to get here with an empty set `s`?
If so, might want to either add another error case for when we don't have a match, or update the `TermAmbiguous` error message to handle an empty set.